### PR TITLE
added indexing to handle tuple

### PR DESCRIPTION
**Pull request by ZohebShaikh**
closes https://github.com/DiamondLightSource/blueapi/issues/444

With PR for dodal at https://github.com/DiamondLightSource/dodal/pull/550

### CODE REVIEW

**Comment by ZohebShaikh** [comment]
The tests are failing now because it is testing it with the old version of dodal

File: src/blueapi/core/context.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def with_device_module(self, module: ModuleType) -> None:
     def with_dodal_module(self, module: ModuleType, **kwargs) -> None:
         from dodal.utils import make_all_devices
 
-        for device in make_all_devices(module, **kwargs).values():
+        for device in make_all_devices(module, **kwargs)[0].values():
             self.device(device)
```
Comment (callumforrester) [inline src/blueapi/core/context.py:110]: I think we want the issues logged

```suggestion
        devices, exceptions = make_all_devices(module, **kwargs)
        for device in devices.values():
            self.device(device)
        
        num_failed = len(exceptions)
        LOGGER.warning(f"{num_failed} devices failed to connect: {exceptions}")
```

**Review by callumforrester — CHANGES_REQUESTED** [review]
Tests would be useful, even they are just copy/pastes of the dodal tests for now, we can work on consolidation later.

File: src/blueapi/core/context.py [hunk 2]
```diff
@@ -106,9 +106,15 @@ def with_device_module(self, module: ModuleType) -> None:
     def with_dodal_module(self, module: ModuleType, **kwargs) -> None:
         from dodal.utils import make_all_devices
 
-        for device in make_all_devices(module, **kwargs).values():
+        devices, exceptions = make_all_devices(module, **kwargs)
+
+        for device in devices.values():
             self.device(device)
 
+        num_failed = len(exceptions)
```
Comment (DiamondJoseph) [inline src/blueapi/core/context.py:114]: Should we register devices that had exceptions?

Comment (DiamondJoseph) [reply src/blueapi/core/context.py:114]: Has the signature of make_all_devices changed? Seems it used to return `dict[str, Device]` but now returns `tuple[dict[str, Device], dict[str, Exception]]`?

Comment (callumforrester) [reply src/blueapi/core/context.py:114]: It changes in the corresponding dodal PR, see description: https://github.com/DiamondLightSource/dodal/pull/550

**Comment by callumforrester** [comment]
Tests will pass following a dodal release

File: src/blueapi/core/context.py [hunk 3]
```diff
@@ -104,11 +103,17 @@ def with_device_module(self, module: ModuleType) -> None:
...
+        devices, exceptions = make_all_devices(module, **kwargs)
 
-        for device in make_all_devices(module, **kwargs).values():
+        for device in devices.values():
             self.device(device)
 
+        # If exceptions have occurred, we log them but we do not make blueapi
+        # fall over
+        if len(exceptions) > 0:
+            LOGGER.warning(f"{len(exceptions)} device(s) failed to connect")
```
Comment (DiamondJoseph) [inline src/blueapi/core/context.py:114]:
```suggestion
            LOGGER.warning(f"{len(exceptions)} exceptions while attempting to connect ")
```

**Review by DiamondJoseph — APPROVED** [review]

**Pull request merged**